File: ExpressionTreeTestObjects/CSCompiler/NewArray.cs
using System;
using System.Collections.Generic;
using System.Linq.Expressions;
using static ExpressionTreeTestObjects.Functions;
using static ExpressionTreeTestObjects.Categories;

namespace ExpressionTreeTestObjects {
    partial class CSCompiler {

        [Category(NewArray)]
        internal static readonly Expression SingleDimensionInit = Expr(() => new string[] { "" });

        [Category(NewArray)]
        internal static readonly Expression SingleDimensionInitExplicitType = Expr(() => new object[] { "" });

        [Category(NewArray)]
        internal static readonly Expression SingleDimensionWithBounds = Expr(() => new string[5]);

        [Category(NewArray)]
        internal static readonly Expression MultidimensionWithBounds = Expr(() => new string[2, 3]);

        [Category(NewArray)]
        internal static readonly Expression JaggedWithElementsImplicitType = Expr(() => new string[][] {
            new [] {"ab","cd" },
            new [] {"ef","gh"}
        });

        [Category(NewArray)]
        internal static readonly Expression JaggedWithElementsExplicitType = Expr(() => new object[][] {
            new [] {"ab","cd" },
            new [] {"ef","gh"}
        });

        [Category(NewArray)]
        internal static readonly Expression JaggedWithBounds = Expr(() => new string[5][]);

        [Category(NewArray)]
        internal static readonly Expression ArrayOfMultidimensionalArray = Expr(() => new string[5][,]);

        [Category(NewArray)]
        internal static readonly Expression MultidimensionalArrayOfArray = Expr(() => new string[3, 2][]);
    }
}

File: ExpressionTreeTestObjects.VB/VBCompiler/NewArray.vb
Partial Friend Module VBCompiler

    <Category(NewArray)>
    Friend SingleDimensionInit As Expression = Expr(Function() New String() {""})

    <Category(NewArray)>
    Friend SingleDimensionInitExplicitType As Expression = Expr(Function() New Object() {""})

    <Category(NewArray)>
    Friend SingleDimensionWithBounds As Expression = Expr(Function() New String(4) {})

    <Category(NewArray)>
    Friend MultidimensionWithBounds As Expression = Expr(Function() New String(1, 2) {})

    <Category(NewArray)>
    Friend JaggedWithElementsImplicitType As Expression = Expr(Function() {
            ({"ab", "cd"}),
            ({"ef", "gh"})
        })

    ' for jagged array literals, the inner literals need to be wrapped in parentheses
    ' we're checking that this is only done for literals
    <Category(NewArray)>
    Friend JaggedWithElementsImplicitTypeInnerNonLiteral As Expression = IIFE(Function()
                                                                                  Dim arr1 = New String() {"ab", "cd"}
                                                                                  Dim arr2 = New String() {"ef", "gh"}
                                                                                  Return Expr(Function() {
                                                                                                      arr1,
                                                                                                      arr2
                                                                                        })
                                                                              End Function)

    <Category(NewArray)>
    Friend JaggedWithElementsExplicitType As Expression = Expr(Function() New Object()() {
            ({"ab", "cd"}),
            ({"ef", "gh"})
        })

    <Category(NewArray)>
    Friend JaggedWithBounds As Expression = Expr(Function() New String(4)() {})

    <Category(NewArray)>
    Friend ArrayOfMultidimensionalArray As Expression = Expr(Function() New String(4)(,) {})

    <Category(NewArray)>
    Friend MultidimensionalArrayOfArray As Expression = Expr(Function() New String(2, 1)() {})
End Module